intake_xarray/image.py

```diff
@@ -0,0 +1,147 @@
+import xarray as xr
+from .base import XarraySource
+
+
+def add_leading_dimension(x):
+    return x[None, ...]
+
+
+def dask_imread(filename, imread=None, preprocess=None):
+    """ Read a stack of images into a dask array
+    Parameters
+    ----------
+    filename: string, iter
+        A globstring like 'myfile.*.png', a list of files, or anything
+        that can go into imread.
+    imread: function (optional)
+        Optionally provide custom imread function.
+        Function should expect a filename and produce a numpy array.
+        Defaults to ``skimage.io.imread``.
+    preprocess: function (optional)
+        Optionally provide custom function to preprocess the image.
+        Function should expect a numpy array for a single image.
+    Examples
+    --------
+    >>> from dask.array.image import imread
+    >>> im = imread('2015-*-*.png')  # doctest: +SKIP
+    >>> im.shape  # doctest: +SKIP
+    (365, 1000, 1000, 3)
+    Returns
+    -------
+    Dask array of all images stacked along the first dimension.  All images
+    will be treated as individual chunks
+    """
+    import os
+    from glob import glob
+
+    try:
+        from skimage.io import imread as sk_imread
+    except (AttributeError, ImportError):
+        pass
+
+    from dask.array import Array
+    from dask.base import tokenize
+
+    imread = imread or sk_imread
+    name = None
+
+    if isinstance(filename, list):
+        filenames = filename
+    elif '*' in filename:
+        filenames = sorted(glob(filename))
+    elif os.path.isfile(filename):
+        filenames = [filename]
+    else:
+        filenames = [filename]
+        name = 'imread-%s' % tokenize(filenames)
+
+    if not filenames:
+        raise ValueError("No files found under name %s" % filename)
+
+    name = name or 'imread-%s' % tokenize(filenames, map(os.path.getmtime, filenames))
+
+    sample = imread(filenames[0])
+    if preprocess:
+        sample = preprocess(sample)
+
+    keys = [(name, i) + (0,) * len(sample.shape) for i in range(len(filenames))]
+    if preprocess:
+        values = [(add_leading_dimension, (preprocess, (imread, fn)))
+                  for fn in filenames]
+    else:
+        values = [(add_leading_dimension, (imread, fn))
+                  for fn in filenames]
+    dsk = dict(zip(keys, values))
+
+    chunks = ((1, ) * len(filenames), ) + tuple((d, ) for d in sample.shape)
+
+    return Array(dsk, name, chunks, sample.dtype)
+
+
+class ImageSource(XarraySource):
+    """Open a xarray dataset from image files.
+
+    This creates an xarray.DataArray or an xarray.Dataset.
+
+    See http://scikit-image.org/docs/dev/api/skimage.io.html#skimage.io.imread
+    for the file formats supported, and
+     http://docs.dask.org/en/latest/array-api.html#dask.array.image.imread
+    for possible extra arguments.
+
+    NOTE: Although ``skimage.io.imread`` is used by default, any reader function which
```

Comment:
Should somewhere give an example of how that works.

Reply:
(or just remove the capability??)

Reply:
I think if it's that is importable, it's easy to do: `!!python/name:mymodule.process`. Could be seen as a future enhancement?

Reply:
That seems reasonable.